src/two_way_peekable.rs

```diff
@@ -0,0 +1,213 @@
+pub trait TwoWayIterator: Iterator {
+    fn prev(&mut self) -> Option<Self::Item>;
+
+    fn two_way_peekable(self) -> TwoWayPeekable<Self>
+    where
+        Self: Sized,
+        Self::Item: Copy,
+    {
+        TwoWayPeekable {
+            itr: self,
+            peeked: Peeked::None,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Peeked<T> {
+    None,
+    Prev(Option<T>), // remember peeked value even if it was none
+    Next(Option<T>),
+}
+
+pub struct TwoWayPeekable<I>
+where
+    I: TwoWayIterator,
+    I::Item: Copy,
+{
+    itr: I,
+    peeked: Peeked<I::Item>,
+}
```

Comment:
If we're going to the trouble of storing peek values (rather than the simple "advance and then de-advance" approach discussed in #82), then instead of a single `peeked` field, I think we want both `peeked_prev` and `peeked_next`, both of which would be simple `Option`s and would be maintained as the iterator progresses.  And then the `Peeked` enum is unnecessary.

Reply:
I'm not sure I follow.  I could be missing something, but isn't the idea that `prev/next_peeked` would simply store the value that `peek_prev()` and `peek_next()` would return?  Or for that matter, what `prev()` and `next()` would return as well (since the only difference between them and the `peek_*` variants is whether the iterator gets advanced or not).

In other words, if we have the following rope, with the char iterator position represented with a vertical bar:

```
ab|cd
```

Then `prev_peeked` would contain `Some('b')` and `next_peeked` would contain `Some('c')`.

Reply:
It's not caching what was returned by `.next()`, it is caching what was returned by `.peek_next()`, so that when we call `.next()`, we know that iterator is already advanced forward by the previous `.peek_next()` call, so instead of advancing it one more time, we returned the cached value, and set it to `Peeked::None`, so that next `.next()` call will just advance the iterator as normal. So if `peeked == Peeked::Next('c')`, it meens we already peeked at 'c' and iterator is now past it. But we still pretend that it is before it. This is the same way `Peekable` is implemented in std. Making it the way you described it would make it a lot more complicated, since we would need to keep track of where the iterator is after we moved it around for peeking.